cpp/loader/multi_gpu_loader.cc

```diff
@@ -0,0 +1,158 @@
+                                    Module mod) {
+  ObjectPtr<MultiGPULoaderObj> n = make_object<MultiGPULoaderObj>();
+  n->metadata_ = NDArrayCacheMetadata::LoadFromStr(metadata, path_to_metadata);
+  n->param_info_map_.clear();
+  n->metadata_str_ = metadata;
+  std::filesystem::path fs_metadata_path = path_to_metadata;
+  n->model_path_ = fs_metadata_path.parent_path().string();
+
+  n->model_metadata_ = ModelMetadata::FromModule(mod);
+  {
+    TypedPackedFunc<PackedFunc(const std::string&)> get_global_func =
+        [](const std::string& name) -> PackedFunc {
+      const auto* f = tvm::runtime::Registry::Get(name);
+      CHECK(f != nullptr) << "ValueError: Cannot find env function " << name;
+      return *f;
+    };
+    n->fload_params = get_global_func("vm.builtin.ndarray_cache.load_from_record");
+    n->fremove_params = get_global_func("vm.builtin.ndarray_cache.remove");
+    n->fget_params = get_global_func("vm.builtin.ndarray_cache.get");
+  }
+  for (ModelMetadata::Param param : n->model_metadata_.params) {
+    n->param_info_map_[param.name] = param;
+    for (const ModelMetadata::Param::Preproc& preproc : param.preprocs) {
+      const std::string& func_name = preproc.func_name;
+      if (PackedFunc f = mod.defined() ? mod->GetFunction(func_name, true) : nullptr;
+          f != nullptr) {
+        n->preproc_funcs_[func_name] = f;
+      } else if (const PackedFunc* f = tvm::runtime::Registry::Get(func_name)) {
+        n->preproc_funcs_[func_name] = *f;
+      } else {
+        LOG(FATAL) << "ValueError: Undefined function: " << func_name;
+      }
+    }
+  }
+
+  return ObjectRef(std::move(n));
+}
+
+NDArray MultiGPULoaderObj::ApplyPreproc(const ModelMetadata::Param::Preproc& preproc,
+                                        const NDArray& param) const {
+  Device device = param->device;
+  NDArray o = NDArray::Empty(preproc.out_shape, preproc.out_dtype, device);
+  PackedFunc f = this->preproc_funcs_.at(preproc.func_name);
+  int n = static_cast<int>(preproc.params.size());
+  std::vector<TVMValue> tvm_args(n + 2);
+  std::vector<int> type_codes(n + 2);
+  TVMArgsSetter setter(tvm_args.data(), type_codes.data());
+  const DLTensor* w_in = param.operator->();
+  const DLTensor* w_out = o.operator->();
+  setter(0, const_cast<DLTensor*>(w_in));
+  for (int i = 0; i < n; ++i) {
+    setter(i + 1, preproc.params[i]);
+  }
+  setter(n + 1, const_cast<DLTensor*>(w_out));
+  TVMRetValue rv;
+  f.CallPacked(TVMArgs(tvm_args.data(), type_codes.data(), n + 2), &rv);
+  return o;
+}
+
+std::unordered_map<std::string, NDArray> MultiGPULoaderObj::LoadFromRecord(int index) const {
+  std::unordered_map<std::string, NDArray> loaded_params;
+  DiscoWorker* worker = DiscoWorker::ThreadLocal();
+  Device device = worker->default_device;
+  const NDArrayCacheMetadata::FileRecord file_record = metadata_.records.at(index);
+  Array<String> loaded_param_names =
```

Comment:
we already have an implementation from https://github.com/apache/tvm/blob/unity/src/runtime/disco/loader.cc. Why do you re-implement a loading logic?